src/hotspot/share/utilities/intrusiveList.hpp

```diff
@@ -0,0 +1,1588 @@
+
+  /** Reverse iterator type. */
+  using reverse_iterator =
+    Impl::IteratorImpl<T, get_entry, false>;
+
+  /** Reverse iterator type with const elements. */
+  using const_reverse_iterator =
+    Impl::IteratorImpl<std::add_const_t<T>, get_entry, false>;
+
+  /** Make an empty list. */
+  IntrusiveList() : _impl() {}
+
+  /**
+   * Destroy the list.
+   *
+   * precondition: empty()
+   */
+  ~IntrusiveList() = default;
+
+  NONCOPYABLE(IntrusiveList);
+
+  /**
+   * Inserts value at the front of the list.  Does not affect the
+   * validity of iterators or element references for this list.
+   *
+   * precondition: value must not already be in a list using the same entry.
+   * complexity: constant.
+   */
+  void push_front(reference value) {
+    insert(begin(), value);
+  }
+
+  /**
+   * Inserts value at the back of the list.  Does not affect the
+   * validity of iterators or element references for this list.
+   *
+   * precondition: value must not already be in a list using the same entry.
+   * complexity: constant.
+   */
+  void push_back(reference value) {
+    insert(end(), value);
+  }
+
+  /**
+   * Removes the front element from the list, and applies the
+   * disposer, if any, to the removed element.  The list may not be in
+   * a consistent state when the disposer is called.  Invalidates
+   * iterators for the removed element.
+   *
+   * precondition: !empty()
+   * complexity: constant.
+   */
+  void pop_front() {
+    pop_front_and_dispose(NopDisposer());
+  }
+
+  template<typename Disposer>
+  void pop_front_and_dispose(Disposer disposer) {
+    erase_and_dispose(begin(), disposer);
+  }
+
+  /**
+   * Removes the back element from the list, and applies the disposer,
+   * if any, to the removed element.  The list may not be in a
+   * consistent state when the disposer is called.  Invalidates
+   * iterators for the removed element.
+   *
+   * precondition: !empty()
+   * complexity: constant.
+   */
+  void pop_back() {
+    pop_back_and_dispose(NopDisposer());
+  }
+
+  template<typename Disposer>
+  void pop_back_and_dispose(Disposer disposer) {
+    erase_and_dispose(rbegin(), disposer);
+  }
+
+  /**
+   * Returns a [const_]reference to the front element of the list.
+   *
+   * precondition: !empty()
+   * complexity: constant.
+   */
+  reference front() { return *begin(); }
+  const_reference front() const { return *begin(); }
+
+  /**
+   * Returns a [const_]reference to the back element of the list.
+   *
+   * precondition: !empty()
+   * complexity: constant.
+   */
+  reference back() { return *rbegin(); }
+  const_reference back() const { return *rbegin(); }
+
+  /**
+   * Returns a [const_]reference to the n'th element of the list.
+   *
+   * precondition: n < length()
+   * complexity: O(length())
+   */
+  reference operator[](size_type n) {
+    return nth_element(begin(), end(), n);
+  }
+
+  const_reference operator[](size_type n) const {
+    return nth_element(cbegin(), cend(), n);
+  }
```

Comment:
Do we need these operator[]'s?  Neither std::list nor boost::intrusive::list have such, and I don't think any of the
existing intrusive lists in HotSpot have such either.  Maybe there's no real use-case.  Removal would save
25-30 lines.

Reply:
If we cannot find a  real use-case for HotSpot, then maybe we shouldn't include them.

Reply:
I agree with Ivan.

Reply:
OK, removed.  We can always add it back if we find a use-case.